airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -86,151 +79,112 @@ def name(self, value: str) -> None:
         """
-        return self.requester.interpret_response_status(response).error_message
+        mapping = method(**kwargs) or {}
+        keys = set(mapping.keys()) if not isinstance(mapping, str) else set()
+        return mapping, keys
 
     def _get_request_options(
         self,
+        stream_state: Optional[StreamData],
         stream_slice: Optional[StreamSlice],
         next_page_token: Optional[Mapping[str, Any]],
-        requester_method: Callable[..., Mapping[str, Any]],
-        paginator_method: Callable[..., Mapping[str, Any]],
-        stream_slicer_method: Callable[..., Mapping[str, Any]],
-        auth_options_method: Callable[..., Mapping[str, Any]],
-    ) -> MutableMapping[str, Any]:
+        paginator_method: Callable[..., Optional[Union[Mapping[str, Any], str]]],
+        stream_slicer_method: Callable[..., Optional[Union[Mapping[str, Any], str]]],
+    ) -> Union[Mapping[str, Any], str]:
         """
-        Get the request_option from the requester and from the paginator
+        Get the request_option from the requester, the authenticator and extra_options passed in.
         Raise a ValueError if there's a key collision
         Returned merged mapping otherwise
-        :param stream_slice:
-        :param next_page_token:
-        :param requester_method:
-        :param paginator_method:
-        :return:
         """
-
-        # FIXME we should eventually remove the usage of stream_state as part of the interpolation
-        requester_mapping = requester_method(stream_state=self.state, stream_slice=stream_slice, next_page_token=next_page_token)
-        requester_mapping_keys = set(requester_mapping.keys())
-        paginator_mapping = paginator_method(stream_state=self.state, stream_slice=stream_slice, next_page_token=next_page_token)
-        paginator_mapping_keys = set(paginator_mapping.keys())
-        stream_slicer_mapping = stream_slicer_method(stream_slice=stream_slice)
-        stream_slicer_mapping_keys = set(stream_slicer_mapping.keys())
-        auth_options_mapping = auth_options_method()
-        auth_options_mapping_keys = set(auth_options_mapping.keys())
-
-        intersection = (
-            (requester_mapping_keys & paginator_mapping_keys)
-            | (requester_mapping_keys & stream_slicer_mapping_keys)
-            | (paginator_mapping_keys & stream_slicer_mapping_keys)
-            | (requester_mapping_keys & auth_options_mapping_keys)
-            | (paginator_mapping_keys & auth_options_mapping_keys)
-            | (stream_slicer_mapping_keys & auth_options_mapping_keys)
+        paginator_mapping, paginator_keys = self._get_mapping(
+            paginator_method, stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token
+        )
+        stream_slicer_mapping, stream_slicer_keys = self._get_mapping(
+            stream_slicer_method, stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token
```

Comment:
I don't think `stream_state` is guaranteed to be equivalent to the `self.state` we used to pass.

If the cursor is a `PerPartitionCursor`, self.state is the full state while stream_state will be scoped to a single partition.

This might be fine in practice since I wouldn't expect many connectors to reference the stream state directly, but I think this would be a breaking change

Reply:
That's a valid point, however I'm a little confused why we even need
```
        if self.cursor and hasattr(self.cursor, "select_state"):  # type: ignore
            slice_state = self.cursor.select_state(stream_slice)  # type: ignore
        elif self.cursor:
            slice_state = self.cursor.get_stream_state()
        else:
            slice_state = {}
```

then in the first place on master - from what I can tell all the things `self._read_pages(self.parse_records, stream_slice, slice_state)` is calling is within the simple retriever and will ignore `slice_state` eventually and use `self.state`:
* _read_pages calls _fetch_next_page and the record generator function which is parse_records
   *  _fetch_next_page calls request_headers, path, request_params, request_body_json, request_body_data, but all of them are implemented on the simple retriever to ignore the passed in stream_state
   * parse_records calls parse_response
      *  parse_response calls select_records, but it's using self.state instead of the passed in state

In sum, the calculated slice_state is never accessed from what I can tell. Can we remove this whole thing? Or did I miss something here?

Reply:
I honestly don’t have a strong opinion on this… For me, the question is: what is the state interpolation interface we want to offer to the user? My guess is that we don’t want to user to have to find the state in an array of states when the state is per partition so I wouldn’t pass `self.state` and keep the hack. However in practice, I don’t think we have a stream with state per partition that use state interpolation.

Do we still intend to remove the state from the interpolation altogether?

Reply:
Thanks @maxi297 . As it seems like it's actually not used right now and the current behavior is to pass down the full state, let's just keep that for now and split this decision out.

I removed the slice state calculation for now

Reply:
> Do we still intend to remove the state from the interpolation altogether?
Yes